arangod/Graph/Enumerators/YenEnumerator.h

```diff
@@ -0,0 +1,275 @@
+    // allocation, since it only consists of two uint64_t values. This is
+    // why we only act on coordinators here:
+    if (!ServerState::instance()->isCoordinator()) {
+      return item;  // A copy, but this is cheap!
+    }
+    uint8_t const* p = item.vpack();
+    VPackSlice v{p};
+    velocypack::ValueLength s = v.byteSize();
+    auto q = (uint8_t const*)(makeLocalCopy((char const*)p, (size_t)s));
+    return EdgeDocumentToken(VPackSlice{q});
+  }
+
+ private:
+  char* makeLocalCopy(char const* p, size_t s) {
+    auto* batch = &_data.back();
+    if (s > batch->buffer.size() - batch->nextFree) {
+      if (s <= BATCH_SIZE) {
+        _data.emplace_back();
+        _totalSize += BATCH_SIZE;
+        _resourceMonitor.increaseMemoryUsage(BATCH_SIZE);
+      } else {
+        _data.emplace_back(s);
+        _totalSize += s;
+        _resourceMonitor.increaseMemoryUsage(s);
+      }
+      batch = &_data.back();
+    }
+    char* place = &batch->buffer[batch->nextFree];
+    batch->nextFree += s;
+    memcpy(place, p, s);
+    return place;
+  }
+};
+
+class PathValidatorOptions;
+struct TwoSidedEnumeratorOptions;
+
+template<class ProviderType, class Step>
+class PathResult;
+
+// This is a relatively straightforward implementation of Yen's algorithm
+// as described here: https://en.wikipedia.org/wiki/Yen%27s_algorithm
+// It uses single shortest path algorithms by means of a subobject of
+// type `TwoSidedEnumerator` (in the unweighted case) and
+// `WeightedTwoSidedEnumerator` (in the weighted case).
+// The subtlety lies in the template magic going on all around this.
+// The `ProviderType` is used to get actual graph data like neighbours
+// out of the system. Essentially, there can be a `SingleServerProvider`
+// or a `ClusterProvider`, and it can be wrapped with a tracer wrapper or
+// not. This provider is just handed on to the `EnumeratorType`, which is
+// essentially an instance of the `TwoSidedEnumerator` or the
+// `WeightedTwoSidedEnumerator`. However, we have to be able to forbid
+// some vertices and edges. This is handled by putting a wrapper type
+// around the PathValidator, we use the `PathValidatorTabooWrapper` for
+// this. Therefore, the implementation can be rather compact here and
+// uses lots of other functionality to do the heavy lifting.
+// Note that some effort is necessary for memory management. Since we
+// repeatedly call ShortestPath we must constantly reset our subobject,
+// which means that we run into the problem that old references to vertex
+// and edge IDs used in previous path results become invalid. Therefore,
+// we have to copy all these IDs to our own managed memory area.
+// This emplate is instantiated in 8 different ways across the following
+// dimensions:
+//   SingleServerProvider/ClusterProvider
+//   Unweighted/weighted
+//   non-tracing/tracing
+
+template<class ProviderType, class EnumeratorType, bool IsWeighted>
+class YenEnumerator {
```

Comment:
To be honest, I do not know, but I do not think so. Neither `TwoSidedEnumerator` nor `WeightedTwoSidedEnumerator` implement `TraversalEnumerator`. I was not even aware of the existence of `TraversalEnumerator` and there is no explanation in the source code as to what purpose it has or what it abstracts. It seems to only be used in the `TraverserExecutor`, which we do not use here. For k-shortest-paths and all-shortest-paths and k-paths we use the `EnumeratePathsExecutor`. Therefore, I would rather not visit this.

As to the signature of the `getNextPath`. I closely followed what `TwoSidedEnumerator` and `WeightedTwoSidedEnumerator` do, and we have to drop in for them for the `EnumeratePathsExecutor`. So I see little potential to change this, unless we would like to do large scale refactorings here.

Reply:
Ok, I'm only asking because the only code location where `getNextPath()` is called (as far as I see), is inside TraversalExecutor::doOutput:
```
  auto currentPath = traversalEnumerator()->getNextPath();
```
where `traversalEnumerator()` gives a `TraversalEnumerator*`. The `TraversalExecutor` also calls the other functions, e.g. `isDone()` or `skipPath()`. I guessed that this is location where the specific enumerator is called, e.g. the YenEnumerator, which therefore should implement the TraveralEnumerator interface. But now I'm also confused, the TraversalEnumerator is defined only in `OneSidedEnumeratorInterface.cpp`, suggesting that this something specific to the one sided enumerator, and has a lot of additional functions.
So I guess I don't yet see the full picture, let's keep it as it is for now.